source/d3d11/draw_call_tracker.cpp

```diff
@@ -1,105 +1,168 @@
+		_global_counter.drawcalls += 1;
 
-		const auto counters = _counters_per_used_depthstencil.find(depthstencil);
+		ID3D11RenderTargetView *rendertargets[D3D11_SIMULTANEOUS_RENDER_TARGET_COUNT] = { nullptr };
+		ID3D11DepthStencilView *depthstencil = nullptr;
 
-		if (counters != _counters_per_used_depthstencil.end())
+		context->OMGetRenderTargetsAndUnorderedAccessViews(D3D11_SIMULTANEOUS_RENDER_TARGET_COUNT, rendertargets, &depthstencil, 0, 0, nullptr);
+
+		if (depthstencil == nullptr) {
+			// this is a draw call with no depthstencil
+			return;
+		}
+
+		const auto intermediate_snapshot = _counters_per_used_depthstencil.find(depthstencil);
+
+		if (intermediate_snapshot != _counters_per_used_depthstencil.end())
 		{
-			counters->second.vertices += vertices;
-			counters->second.drawcalls += drawcalls;
+			intermediate_snapshot->second.counter.vertices += vertices;
+			intermediate_snapshot->second.counter.drawcalls += 1;
+
+			// Find the render targets, if they exist, and update their counts
+			for (unsigned int i = 0; i < D3D11_SIMULTANEOUS_RENDER_TARGET_COUNT; i++) {
+				if (rendertargets[i] != nullptr) {
+					auto it = intermediate_snapshot->second.additional_views.find(rendertargets[i]);
+					if (it != intermediate_snapshot->second.additional_views.end()) {
+						it->second.vertices += vertices;
+						it->second.drawcalls += 1;
+					}
+					else {
+						// This shouldn't happen - it means somehow someone has called "draw" with a rendertarget without calling track_rendertargets first
+						LOG(ERROR) << "Draw has been called on an untracked rendertarget.";
+					}
+				}
+			}
+		}
+
+		// Capture constant buffers that are used when depthstencils are drawn
+		ID3D11Buffer *vscbuffers[D3D11_COMMONSHADER_CONSTANT_BUFFER_API_SLOT_COUNT] = { nullptr };
+		context->VSGetConstantBuffers(0, D3D11_COMMONSHADER_CONSTANT_BUFFER_API_SLOT_COUNT, vscbuffers);
+
+		for (unsigned int i = 0; i < D3D11_COMMONSHADER_CONSTANT_BUFFER_API_SLOT_COUNT; i++) {
+			// uses the default drawcalls = 0 the first time around.
+			if (vscbuffers[i] != nullptr) {
+				_counters_per_constant_buffer[vscbuffers[i]].vertexshaders += 1;
+			}
+		}
+
+		ID3D11Buffer *pscbuffers[D3D11_COMMONSHADER_CONSTANT_BUFFER_API_SLOT_COUNT] = { nullptr };
+		context->PSGetConstantBuffers(0, D3D11_COMMONSHADER_CONSTANT_BUFFER_API_SLOT_COUNT, pscbuffers);
+
+		for (unsigned int i = 0; i < D3D11_COMMONSHADER_CONSTANT_BUFFER_API_SLOT_COUNT; i++) {
+			// uses the default drawcalls = 0 the first time around.
+			if (pscbuffers[i] != nullptr) {
+				_counters_per_constant_buffer[pscbuffers[i]].pixelshaders += 1;
+			}
+		}
+
+	}
+
+	void draw_call_tracker::on_map(ID3D11DeviceContext *context, ID3D11Resource *pResource) {
+
+		D3D11_RESOURCE_DIMENSION dim;
+		pResource->GetType(&dim);
+
+		if (dim == D3D11_RESOURCE_DIMENSION::D3D11_RESOURCE_DIMENSION_BUFFER) {
+			_counters_per_constant_buffer[reinterpret_cast<ID3D11Buffer*>(pResource)].mapped += 1;
```

Comment:
This should be a `static_cast`, not a `reinterpret_cast`. Also, `D3D11_RESOURCE_DIMENSION::D3D11_RESOURCE_DIMENSION_BUFFER` is doubled, just `D3D11_RESOURCE_DIMENSION_BUFFER` is enough and doesn't lose any information.